api/client/client.go

```diff
@@ -54,48 +52,118 @@ func init() {
 // Client is a gRPC Client that connects to a teleport auth server through TLS.
 type Client struct {
 	c Config
+
+	// tlsConfig is the *tls.Config for a successfully connected client.
+	tlsConfig *tls.Config
+
+	// dialer is the ContextDialer for a successfully connected client.
+	dialer ContextDialer
+
 	// grpc is the gRPC client specification for the auth server.
 	grpc proto.AuthServiceClient
+
 	// conn is a grpc connection to the auth server.
 	conn *grpc.ClientConn
+
 	// closedFlag is set to indicate that the services are closed.
 	closedFlag int32
 }
 
-// NewClient returns a new auth client that uses mutual TLS authentication and
-// connects to the remote server using the Dialer or Addrs given in Config.
-func NewClient(cfg Config) (*Client, error) {
+// New creates a new API client with a connection to a Teleport server.
+func New(ctx context.Context, cfg Config) (*Client, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	c := &Client{c: cfg}
-	dialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+	if err := c.connect(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return c, nil
+}
+
+// getDialer builds a grpc dialer for the client from a ContextDialer.
+// The ContextDialer is chosen from available options, preferring one from
+// credentials, then from configuration, and lastly from addresses.
+func (c *Client) setDialer(creds Credentials) error {
+	var err error
+	if c.dialer, err = creds.Dialer(); err == nil {
```

Comment:
Can the client be shared by multiple goroutines? If so, the dialer field will be subject to a race condition.

Reply:
`connect` is only called in initialization, so it could not happen in multiple go routines. If/when we add re loading I can add synchronization logic.